javascript/ql/lib/semmle/javascript/frameworks/Knex.qll

```diff
@@ -46,17 +46,38 @@ module Knex {
     RawKnexSqlString() { this = any(RawKnexCall call).getArgument(0).asExpr() }
   }
 
-  /** A call that triggers a SQL query submission. */
-  private class KnexDatabaseAccess extends DatabaseAccess {
-    KnexDatabaseAccess() {
+  /** A call that triggers a SQL query submission by calling then/stream/asCallback. */
+  private class KnexDatabaseCallback extends DatabaseAccess, DataFlow::CallNode {
+    KnexDatabaseCallback() {
       this = knexObject().getMember(["then", "stream", "asCallback"]).getACall()
+    }
+
+    override DataFlow::Node getAResult() { 
+      this = knexObject().getMember("then").getACall() and
+        result = this.getCallback(0).getParameter(0)
       or
-      exists(AwaitExpr await |
-        this = await.flow() and
-        await.getOperand() = knexObject().getAUse().asExpr()
-      )
+        this = knexObject().getMember("asCallback").getACall() and
+        result = this.getCallback(0).getParameter(1)
+    }
+
+    override DataFlow::Node getAQueryArgument() {
+      none()
+    }
+  }
+
+  private class KnexDatabaseAwait extends DatabaseAccess, DataFlow::ValueNode {
+    AwaitExpr enclosingAwait;
+    KnexDatabaseAwait() {
+      this = enclosingAwait.flow() and
+      enclosingAwait.getOperand() = knexObject().getAUse().asExpr()
```

Comment:
```suggestion
    KnexDatabaseAwait() {
      exists(AwaitExpr enclosingAwait | this = enclosingAwait.flow() | 
        enclosingAwait.getOperand() = knexObject().getAUse().asExpr()
      )
```

You have a field that is not used outside of the characteristic predicate. The field can be inlined into an `exists(..)`.

Reply:
👍